File: algorithms/03_cpp/numerical/euler_method.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <functional>
#include <iostream>
#include <utility>
#include <vector>

double eulerMethod(const std::function<double(double, double)> &f, double y, double t, double t1, int steps) {
    double h = (t1 - t) / steps;
    for (int i = 0; i < steps; ++i) { y += h * f(t, y); t += h; }
    return y;
}

int main() {
    assert(std::abs(eulerMethod([](double, double y) { return y; }, 1.0, 0.0, 1.0, 1000) - std::exp(1.0)) < 0.01);
    std::cout << "euler method ok\n";
    return 0;
}

File: algorithms/01_python/numerical/euler_method.py
"""Forward Euler method for ODEs (Numerical Recipes 17.1)."""
def euler_method(f, y0, t0, t1, steps=1000):
    h = (t1 - t0) / steps
    y, t = y0, t0
    for _ in range(steps):
        y += h * f(t, y)
        t += h
    return y


if __name__ == "__main__":
    value = euler_method(lambda t, y: y, 1.0, 0.0, 1.0)
    assert abs(value - 2.718281828459045) < 0.01
    print("euler method ok")
